### ISSUE TITLE
Mutiny Session API

### ISSUE THREAD

**Issue by gavinking** [hibernate/hibernate-reactive#94]
Alright, so I've been playing with Mutiny, very superficially, just to get a feel for how the reactive `Session` API would look with this.

So first, it seems that there are sorta two reasonable modes of usage. The first looks like this, and is the the way I think they want me to do it:

```java
openMutinySession()
        .onItem().produceUni( s -> s.persist( new GuineaPig( 10, "Tulip" ) ) )
        .onItem().produceUni( s -> s.flush() )
        .onItem().produceUni( v -> selectNameFromId( 10 ) )
        .onItem().invoke( selectRes -> context.assertEquals( "Tulip", selectRes ) )
```

The second way looks like this:

```java
openMutinySession()
        .flatMap( s -> s.persist( new GuineaPig( 10, "Tulip" ) ) )
        .flatMap( s -> s.flush() )
        .flatMap( v -> selectNameFromId( 10 ) )
        .map( selectRes -> context.assertEquals( "Tulip", selectRes ) )
```

My initial, superficial, impression is that I prefer the look of the second code fragment, because:

- Well, I guess I get bored of typing `.onItem()`.
- I don't love the name `produceUni()`, for a very specific reason: from my point of view of writing business code, there's little difference between `map()` and `flatMap()`, and thus I definitely like the fact that they have similar names.

Purely due to the name, `produceUni()` and `apply()` look like completely different operations, but that's not how I view them in my code. I would have thought that chaining `Uni`s would be like almost the most basic operation in Mutiny, and I would have expected it to have a really snappy name. But OK, I guess `flatMap()` is not *especially* snappy. I guess my real question is why is it `produceUni()` instead of `applyUni()`, which is definitely what I feel like I'm doing in my code?

OK, but fine, so anyway I'm rather happy with `map()` and `flatMap()` except for one little thing: AFAICS, there's no real equivalent to `invoke()` at the level of `map()` and `flatMap()`, which is a problem because sometimes I have `void` functions. If I had one, I think I would be about 98% happy, which is certainly a very high level of happiness for someone so curmudgeonly.

(Thankfully, Mutiny gets the return type of `invoke()` correct, one of my gripes with `CompletionStage` is that `thenAccept()` throws away my stream elements.)

@FroMage @cescoffier reactions?

**Comment by gavinking** [comment]
P.S. I could push some code onto a branch if anyone else wants to try this out.

**Comment by DavideD** [comment]
> P.S. I could push some code onto a branch if anyone else wants to try this out.

Thanks, That would be nice

**Comment by gavinking** [comment]
Here's another pair of examples:

```java
				populateDB()
						.onItem().produceUni( v -> openMutinySession() )
						.onItem().produceUni( session ->
								session.find( GuineaPig.class, 5 )
										.onItem().produceUni( aloi -> session.remove( aloi.get() ) )
										.onItem().produceUni( v -> session.flush() )
										.onItem().produceUni( v -> selectNameFromId( 5 ) )
										.onItem().invoke( ret -> context.assertNull( ret ) ) )
						.convert().toCompletionStage()
```

vs:

```java
				populateDB()
						.flatMap( v -> openMutinySession() )
						.flatMap( session ->
							session.find( GuineaPig.class, 5 )
								.flatMap( aloi -> session.remove( aloi.get() ) )
								.flatMap( v -> session.flush() )
								.flatMap( v -> selectNameFromId( 5 ) )
								.map( ret -> context.assertNull( ret ) ) )
								.convert().toCompletionStage()
````

**Comment by FroMage** [comment]
Well, isn't `map` a shortcut for `onItem().produceUni`? I mean both options are equivalent, no?

**Comment by gavinking** [comment]
> Well, isn't `map` a shortcut for `onItem().produceUni`? I mean both options are equivalent, no?

Isn't that `flatMap()`

**Comment by cescoffier** [comment]
@gavinking 

> produceUni() instead of applyUni(), which is definitely what I feel like I'm doing in my code?

That actually can still be changed. To be honest, I don't like `produceX` either. 

> Well, isn't map a shortcut for onItem().produceUni? I mean both options are equivalent, no?

`map` is `onItem().apply`
`flatMap` is `onItem().produceUni()`

**Comment by cescoffier** [comment]
About `applyX`, I want to rethink this part very soon. As I said, I'm not satisfied either. `applyX` is a good idea. I need to see the results. 

About having shortcut for `onItem().invoke()`, it's mostly a name issue... I don't like the traditional `doOnNext`. Maybe `invokeOnItem() / invokeOnFailure()`, or the opposite `onItemInvoke(...);`, or just `onItem(...)`.

**Comment by gavinking** [comment]
> About having shortcut for `onItem().invoke()`, it's mostly a name issue...

I don't completely *hate* `accept()` but I don't love it either. I will try to think of something better.

**Comment by gavinking** [comment]
Oh, of course, actually the traditional name for this operation is `forEach()` or `each()`. Perhaps that makes sense, given that `Multi` also needs it...

**Comment by gavinking** [comment]
See: <https://github.com/gavinking/hibernate-rx/tree/mutiny>

**Comment by aguibert** [comment]
I tried out the Mutiny Session API proposal locally and it looks good @gavinking 

+1 on a shortcut for `onItem().invoke()`, since that is an extremely common operation (but that's more of a comment for @cescoffier for Mutiny API itself)

The only other comment I had was to make the entry point a bit cleaner (the `openMutinySession()` helper method is pretty verbose), but I assume we will refine API entry points later on since the non-Mutiny entry point still uses `.unwrap( RxSessionFactory.class )`

**Comment by gavinking** [comment]
> The only other comment I had was to make the entry point a bit cleaner 

Yes yes, of course, please ignore the code which constructs a `Mutiny.Session` using `new`. That's of course going to look like `.unwrap(Mutiny.SessionFactory).createMutinySession()` or whatever.

**Comment by aguibert** [comment]
@gavinking are we waiting on anyone else to weigh in on this? If no, want to go ahead and make a PR with your proposed changes in your mutiny branch?

**Comment by gavinking** [comment]
@aguibert It's not ready yet, there's still some stuff to figure out.

**Comment by gavinking** [comment]
@cescoffier under what sort of conditions does a subscription cancellation happen? Does that amount to something that should cause a transaction rollback?

**Comment by gavinking** [comment]
Done, with followup issue #153.

**Comment by cescoffier** [comment]
@gavinking To answer your question about cancellations. Cancellations are events flowing from downstream to upstream. They can be emitted if the consumer (subscriber) cancels the subscription (does not need any more events from upstream), or if something downstream fails and so resources must be cleaned up. 

In your case, a cancellation event would mean a rollback I believe. 

There is an `on().termination()` operator that is called:

* on successful completion
* on failiure
* on cancellation

The parameters let you know in which case you are.

**Comment by gavinking** [comment]
OK, thanks, I will create an issue to make cancellation cause a rollback. Failures already cause rollbacks.

**Comment by aguibert** [comment]
@gavinking I just played around with the new Mutiny.Session API in the Quarkus extension, and I really prefer the original version where `Mutiny.SessionFactory#openSession` returned a `Mutiny.Session` instead of the current `Uni<Mutiny.Session>` (same applies for the `Stage` variant too).

I have two main pain points for wrapping the reactive session in some sort of reactive pipeline object:
1. I often need to compose nested stages to be able to access the session in all sub-stages. The regular stage-to-stage flow works well if you only ever operate on the results of the previous stage, but for more "global" objects like the session, it's handy to be able to use it in a stage, do something else in the next stage, and then access the session again in a later stage.
2. CDI management is not very clear. Before we could simply open the session in the producer and close the session in the disposer. Now, producing the Uni/CS works fine, but disposing is awkward. With Mutiny I tried:
```java
    public void disposeMutinySession(@Disposes Uni<Mutiny.Session> reactiveSession) {
        reactiveSession.on().termination((session, ex, cancelled) -> {
            if (session != null)
                session.close();
        });
    }
```
But the `termination()` block never gets called for a JAX-RS endpoint that produces a simple `Uni<Thing>`. CC @cescoffier in case I'm just mis-using the Mutiny API here.

**Comment by gavinking** [comment]
>  I just played around with the new `Mutiny.Session` API in the Quarkus extension, and I really prefer the original version where `Mutiny.SessionFactory#openSession` returned a `Mutiny.Session` instead of the current `Uni<Mutiny.Session>`

Yes, I agree it was more convenient.

The thing is that I also like the idea that the session lifecycle maps exactly to the lifecycle of the underlying Vert.x `SqlClient`. Opening a `Session` means obtaining a connection, closing the `Session` means closing it.

Now, it's certainly *possible* to go back to opening the connection lazily when it's first used, thus hiding this aspect of its lifecycle inside the `SqlClientPool`, but I could sorta make the same argument that the Vert.x `SqlClient` could do the same thing, but it doesn't :-)

So I'm not sure. I guess I could revert to the previous design if you guys think it's better.

> I often need to compose nested stages to be able to access the session in all sub-stages. The regular stage-to-stage flow works well if you only ever operate on the results of the previous stage, but for more "global" objects like the session, it's handy to be able to use it in a stage, do something else in the next stage, and then access the session again in a later stage.

Well I think you can still *do* that. You just use

    sf.openSession(). thenCompose( s -> {
        //all my code here in a block
    } )

Or even better:

    sf.withSession( s -> {
        //all my code here in a block
    } )

> CDI management is not very clear. Before we could simply open the session in the producer and close the session in the disposer. Now, producing the Uni/CS works fine, but disposing is awkward. 

You would probably want to proxy it. `java.lang.reflect.Proxy` should do the trick.

(Just like you want me to proxy the `SqlConnection`.)

**Comment by aguibert** [comment]
> Now, it's certainly possible to go back to opening the connection lazily when it's first used, thus hiding this aspect of its lifecycle inside the SqlClientPool, 

I think the important thing is that we keep the same cardinality (1 session gets 1 connection). The important bits of the lifecycle still hold up (the connection lifecycle is bound by the session lifecycle), we're just decoupling the starting points of that lifecycle.

> but I could sorta make the same argument that the Vert.x SqlClient could do the same thing, but it doesn't :-)

Agreed, and I've argued (unsuccessfully) against this in Vertx as well.

Regarding stage composition, to illustrate my point currently we'll have to do things like this in a JAX-RS resource:
```java
@Inject
Uni<Mutiny.Session> sessionUni;
// ...

return sessionUni.flatMap(actualSession -> {
  return actualSession.find(....)
            .flatMap(/* do some other stuff */)
            .flatMap(junk -> actualSession.persist(...));
});
```
As opposed to not needing to nest a sub-stage if we can access the session anywhere:
```java
@Inject
Mutiny.Session actualSession;
// ...

return actualSession.find(....)
          .flatMap(/* do some other stuff */)
          .flatMap(junk -> actualSession.persist(...));
```

I'm willing to defer to @cescoffier on this though. To me the simpler way makes more sense, but perhaps the simpler way is bad because it could enable creating multiple streams off of the `actualSession` resulting in side-effect streams 🤷

**Comment by gavinking** [comment]
@aguibert see https://github.com/hibernate/hibernate-reactive/pull/196
